stacks-signer/src/crypto/frost.rs

```diff
@@ -0,0 +1,888 @@
+            let aggregate_nonce = self.compute_aggregate_nonce();
+
+            // check to see if aggregate public key has even y
+            if aggregate_nonce.has_even_y() {
+                info!("Aggregate nonce has even y coord!");
+                info!("Aggregate nonce: {}", aggregate_nonce);
+                self.move_to(State::SigShareRequest)?;
+            } else {
+                warn!("Sign Round #{} Nonce Round #{} Failed: Aggregate nonce does not have even y coord, requesting new nonces.", self.current_sign_id, self.current_sign_nonce_id);
+                self.current_sign_nonce_id += 1;
+                self.move_to(State::NonceRequest)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn request_sig_shares(&mut self) -> Result<Message, Error> {
+        info!(
+            "Sign Round #{} Requesting Signature Shares",
+            self.current_sign_id,
+        );
+        let nonce_responses = (0..self.total_signers)
+            .map(|i| self.public_nonces[&i].clone())
+            .collect::<Vec<NonceResponse>>();
+        let sig_share_request = SignatureShareRequest {
+            dkg_id: self.current_dkg_id,
+            sign_id: self.current_sign_id,
+            correlation_id: 0, // TODO: what is this?
+            nonce_responses,
+            message: self.message.clone(),
+        };
+        let sig_share_request_msg = Message {
+            sig: sig_share_request.sign(&self.message_private_key).expect(""),
+            msg: MessageTypes::SignShareRequest(sig_share_request), // TODO: either SigShare or SignatureShare, SignShare is right out
+        };
+        self.ids_to_await = (0..self.total_signers).collect();
+        self.move_to(State::SigShareGather)?;
+        Ok(sig_share_request_msg)
+    }
+
+    fn gather_sig_shares(&mut self, message: &Message) -> Result<(), Error> {
+        if let MessageTypes::SignShareResponse(sig_share_response) = &message.msg {
+            if sig_share_response.dkg_id != self.current_dkg_id {
+                return Err(Error::BadDkgId(
+                    sig_share_response.dkg_id,
+                    self.current_dkg_id,
+                ));
+            }
+            if sig_share_response.sign_id != self.current_sign_id {
+                return Err(Error::BadSignId(
+                    sig_share_response.sign_id,
+                    self.current_sign_id,
+                ));
+            }
+            self.signature_shares.insert(
+                sig_share_response.signer_id,
+                sig_share_response.signature_shares.clone(),
+            );
+            self.ids_to_await.remove(&sig_share_response.signer_id);
+            info!(
+                "Sign round #{} SignShareResponse from signer #{}. Waiting on {:?}",
+                sig_share_response.sign_id, sig_share_response.signer_id, self.ids_to_await
+            );
+        }
+        if self.ids_to_await.is_empty() {
+            // Calculate the aggregate signature
```

Comment:
Would it make sense to use a draining iterator for the following?  This state won't be used again in this signing round, right?  It would save you some `clone()`s and ensure this signature generation step only happens at most once.